### PR TITLE
♻ Simplify `iterator` and `removeWhere` logic

### PR DESCRIPTION
**Pull request by lig**

### CODE REVIEW

File: lib/ordered_set.dart [hunk 1]
```diff
@@ -72,18 +72,8 @@ class OrderedSet<E> extends IterableMixin<E> implements Iterable<E> {
...
-      _length -= removed.length;
-
-      if (removed.length == es.length) {
-        _backingSet.remove(es);
-      } else {
-        es.removeWhere(test);
-      }
-    }
+    final List<E> _removed = this.where(test).toList();
+    for (final e in _removed) this.remove(e);
```
Comment (lig) [inline lib/ordered_set.dart]: this could use forEach, actualy

**Comment by lig** [comment]
@luanpotter what do you think?

File: lib/ordered_set.dart [hunk 2]
```diff
@@ -72,19 +72,7 @@ class OrderedSet<E> extends IterableMixin<E> implements Iterable<E> {
...
-      _length -= removed.length;
-
-      if (removed.length == es.length) {
-        _backingSet.remove(es);
-      } else {
-        es.removeWhere(test);
-      }
-    }
-    return _removed;
+    return this.where(test).toList()..forEach((e) {this.remove(e);});
```
Comment (luanpotter) [inline lib/ordered_set.dart]: we can use an inline lambda here

```suggestion
    return this.where(test).toList()..forEach((e) => this.remove(e));
```

File: lib/ordered_set.dart [hunk 3]
```diff
@@ -48,7 +48,7 @@ class OrderedSet<E> extends IterableMixin<E> implements Iterable<E> {
 
   @override
   Iterator<E> get iterator {
-    return _OrderedSetIterator<E>(this);
+    return _backingSet.expand<E>((es) => es).iterator;
```
Comment (luanpotter) [inline lib/ordered_set.dart:51]: nice, didn't know about expand!

Comment (luanpotter) [reply lib/ordered_set.dart:51]: will this guarantee the order?

**Comment by luanpotter** [comment]
Nice cleanup! So should we merge this first and then the non-null thing? @lig 
(as this can improve both the current release and the first non-null release)

File: lib/ordered_set.dart (hunk 3 above)
Comment (lig) [reply lib/ordered_set.dart:51]: as far as I see in the code it does almost the same as what `_OrderedSetIterator` did except it is already supporting null safety and it behaves close to the iterator protocol in some cases, i.e. before the first `moveNext` call

**Comment by lig** [comment]
@luanpotter I would like to get this in first and then rebase null safety on it. As far as I see around what others are doing in other projects it is better to keep the null safety code in a separate branch for now and release `nullsafety` suffixed package from that branch.

File: lib/ordered_set.dart (hunk 2 above)
Comment (lig) [reply lib/ordered_set.dart]: Nice catch. How about converting it to oneliner then?:)

**Comment by lig** [comment]
@luanpotter great thanks! Give me some time to rebase #5 then.

**Pull request merged**